src/util/taproot.rs

```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        // key corresponding to the internal key.
+        // Because the tweak is derived as specified in BIP341 (hash output of a function),
+        // this is unlikely to occur (1/2^128) in real life usage, it is safe to unwrap this
+        let parity = output_key
+            .tweak_add_assign(&secp, &tweak)
+            .expect("Cryptographically unreachable code");
+        Self {
+            internal_key: internal_key,
+            merkle_root: merkle_root,
+            output_key_parity: parity,
+            output_key: output_key,
+            script_map: BTreeMap::new(),
+        }
+    }
+
+    /// Obtain the tweak and parity used to compute the output_key
+    pub fn tap_tweak(&self) -> sha256::Hash {
+        compute_tap_tweak(self.merkle_root, self.internal_key)
+    }
+
+    /// Output key(the key used in script pubkey) from Spend data. See also
+    /// [`TaprootSpendInfo::output_key_parity`]
+    pub fn output_key(&self) -> schnorr::PublicKey {
+        self.output_key
+    }
+
+    /// Parity of the output key. See also [`TaprootSpendInfo::output_key`]
+    pub fn output_key_parity(&self) -> bool {
+        self.output_key_parity
+    }
+
+    // Internal function to compute [`TaprootSpendInfo`] from NodeInfo
+    fn from_node_info<C: secp256k1::Verification>(
+        secp: &Secp256k1<C>,
+        internal_key: schnorr::PublicKey,
+        node: NodeInfo,
+    ) -> TaprootSpendInfo {
+        // Create as if it is a key spend path with the given merkle root
+        let mut info = TaprootSpendInfo::new_key_spend(secp, internal_key, Some(node.hash));
+        for leaves in node.leaves {
+            let key = (leaves.script, leaves.ver);
+            let value = leaves.merkle_branch;
+            match info.script_map.get_mut(&key) {
+                Some(set) => {
+                    set.insert(value);
+                    continue; // NLL fix
+                }
+                None => {}
+            }
+            let mut set = BTreeSet::new();
+            set.insert(value);
+            info.script_map.insert(key, set);
+        }
+        info
+    }
+
+    /// Access the internal script map
+    pub fn as_script_map(&self) -> &BTreeMap<(Script, LeafVersion), BTreeSet<TaprootMerkleBranch>> {
+        &self.script_map
+    }
+
+    /// Obtain a [`ControlBlock`] for particular script with the given version.
+    /// Returns [`None`] if the script is not contained in the [`TaprootSpendInfo`]
+    /// If there are multiple ControlBlocks possible, this returns the shortest one.
+    pub fn control_block(&self, script_ver: &(Script, LeafVersion)) -> Option<ControlBlock> {
+        let merkle_branch_set = self.script_map.get(script_ver)?;
+        // Choose the smallest one amongst the multiple script maps
+        let smallest = merkle_branch_set
+            .iter()
+            .min_by(|x, y| x.0.len().cmp(&y.0.len()))
+            .expect("Non-empty iterator");
```

Comment:
This will panic for non-script structures. We may assume we guarantee that the script is always present, but if I will set `x.merkle_root = None` (a public writable field) than there is no script, but the index is still there, and while we do not panic, we will produce incorrect control block. So either all fields must be private (modifiable via functions) - nor we should not assume that we always have a script and get rid of indexes

Reply:
Indeed, changing Merkle root and internal key to non-pub should fix all of this.

Reply:
In which case, I think the expect here would be okay? 

I like having assertions/expect in code because they help reason about algorithmic invariants. If the map has a key the value must have at least one element.

Reply:
Yes, if invariant is guaranteed by the API, `except()` is OK, although I'd like the message to be something like "broken invariant that merkle_branch_set is non-empty".